Saving edits fails due to date parsing problem

2014-09-25 08:33:50,195 [ERROR] django.request: Internal Server Error: /latreemap/features/369463/
Traceback (most recent call last):
  File "/usr/local/otm/env/local/lib/python2.7/site-packages/django/core/handlers/base.py", line 114, in get_response
    response = wrapped_callback(request, _callback_args, *_callback_kwargs)
  File "/usr/local/otm/app/opentreemap/treemap/decorators.py", line 43, in wrapper
    return view_fn(request, instance, _args, *_kwargs)
  File "/usr/local/otm/app/opentreemap/opentreemap/util.py", line 26, in routed
    return kwargs['ELSE'](request, *args2, **kwargs2)
  File "/usr/local/otm/app/opentreemap/treemap/decorators.py", line 253, in wrapper
    return view_fn(request, _args, *_kwargs)
  File "/usr/local/otm/app/opentreemap/treemap/decorators.py", line 208, in newreq
    outp = req_function(request, _args, *_kwargs)
  File "/usr/local/otm/app/opentreemap/treemap/decorators.py", line 191, in newreq
    outp = req_function(request, _args, *_kwargs)
  File "/usr/local/otm/app/opentreemap/treemap/decorators.py", line 298, in wrapper
    return view_fn(request, instance, _args, *_kwargs)
  File "/usr/local/otm/app/opentreemap/opentreemap/util.py", line 24, in routed
    return req_method(request, _args2, *_kwargs2)
  File "/usr/local/otm/app/opentreemap/treemap/views/map_feature.py", line 116, in update_map_feature_detail
    return _request_to_update_map_feature(request, instance, feature)
  File "/usr/local/otm/app/opentreemap/treemap/views/map_feature.py", line 35, in _request_to_update_map_feature
    feature, tree = update_map_feature(request_dict, request.user, feature)
  File "/usr/local/otm/env/local/lib/python2.7/site-packages/django/db/transaction.py", line 339, in inner
    return func(_args, *_kwargs)
  File "/usr/local/otm/app/opentreemap/treemap/views/map_feature.py", line 236, in update_map_feature
    errors.update(save_and_return_errors(tree, user))
  File "/usr/local/otm/app/opentreemap/treemap/views/map_feature.py", line 197, in save_and_return_errors
    thing.save_with_user(user)
  File "/usr/local/otm/app/opentreemap/treemap/models.py", line 769, in save_with_user
    self.full_clean_with_user(user)
  File "/usr/local/otm/app/opentreemap/treemap/audit.py", line 824, in full_clean_with_user
    super(Auditable, self).full_clean(exclude=exclude_fields)
  File "/usr/local/otm/env/local/lib/python2.7/site-packages/django/db/models/base.py", line 928, in full_clean
    self.clean_fields(exclude=exclude)
  File "/usr/local/otm/app/opentreemap/treemap/udf.py", line 1192, in clean_fields
    self.clean_udfs()
  File "/usr/local/otm/app/opentreemap/treemap/udf.py", line 1173, in clean_udfs
    collection_field.clean_collection(data)
  File "/usr/local/otm/app/opentreemap/treemap/udf.py", line 710, in clean_collection
    subfield_val, datatype)
  File "/usr/local/otm/app/opentreemap/treemap/udf.py", line 682, in clean_value
    return datetime.strptime(value.strip(), DATE_FORMAT)
  File "/usr/lib/python2.7/_strptime.py", line 325, in _strptime
    (data_string, format))
ValueError: time data '07/21/2014' does not match format '%Y-%m-%d'

--- a/opentreemap/treemap/js/src/inlineEditForm.js
+++ b/opentreemap/treemap/js/src/inlineEditForm.js
@@ -204,6 +204,7 @@ exports.init = function(options) {
                         .map(function() {
                             return $(this).html();
                         });
+
                 headers = _.compact(headers);
 
                 data[name] =
@@ -214,7 +215,7 @@ exports.init = function(options) {
 
                             rowData = _.object(headers, $tds
                                         .map(function() {
-                                            return $.trim($(this).html());
+                                            return $.trim($(this).attr('data-value'));
                                         }));
                         if (! _.isEmpty(id)) {
                             rowData.id = id;

--- a/opentreemap/treemap/js/src/plot.js
+++ b/opentreemap/treemap/js/src/plot.js
@@ -24,7 +24,7 @@ exports.init = function(options) {
     var udfRowTemplate = _.template(
         '<tr data-value-id="">' +
             '<% _.each(fields, function (field) { %>' +
-            '<td> <%= field %> </td>' +
+            '<td data-value="<%= field.value %>"> <%= field.display %> </td>' +
             '<% }) %>' +
             '</tr>');
 
@@ -34,10 +34,20 @@ exports.init = function(options) {
         var fields = $('table[data-udf-id="' + id + '"] * [data-field-name]').toArray();
 
         var data = _.map(fields, function(field) {
-            if ($(field).attr('data-moment-date-format')) {
-                return moment($(field).datepicker("getDate")).format($(field).attr('data-moment-date-format'));
+            var $field = $(field);
+            if ($field.attr('data-date-display-format')) {
+                var displayFormat = $field.attr('data-date-display-format'),
+                    valueFormat = $field.attr('data-date-serialize-format'),
+                    date = $field.datepicker("getDate");
+                return {
+                    display: moment(date).format(displayFormat),
+                    value: moment(date).format(valueFormat),
+                };
             } else {
-                return $(field).val();
+                return {
+                    value: $field.val(),
+                    display: $field.val()
+                };
             }
         });
 

--- a/opentreemap/treemap/templates/treemap/partials/collection_udf_row.html
+++ b/opentreemap/treemap/templates/treemap/partials/collection_udf_row.html
@@ -6,15 +6,17 @@
    {% endif %}
    >
   {% for field in udf.datatype_dict %}
-  <td>
   {% with field_value=value|get:field.name %}
     {% if field.type == "date" %}
-    {{ field_value|date:request.instance.short_date_format }}
-    {% else %}
-    {{ field_value }}
+    <td data-value="{{ field_value|date:"Y-m-d" }}">
+      {{ field_value|date:request.instance.short_date_format }}
+    </td>
+      {% else %}
+    <td data-value="{{ field_value }}">
+      {{ field_value }}
+    </td>
     {% endif %}
   {% endwith %}
-  </td>
   {% endfor %}
   <td></td>
 </tr>

--- a/opentreemap/treemap/templates/treemap/partials/udf_edit_field.html
+++ b/opentreemap/treemap/templates/treemap/partials/udf_edit_field.html
@@ -1,4 +1,6 @@
 {% load i18n %}
+{% load dateformat %}
+
 {% if field.type == 'choice' %}
 <select class="form-control" data-field-name="{{ field.name }}">
   {% for choice in field.choices %}
@@ -6,8 +8,9 @@
   {% endfor %}
 </select>
 {% elif field.type == 'date' %}
-<input class="form-control" type="text" data-field-name="{{ field.name }}" data-date-format="yyyy-mm-dd"
-       data-moment-date-format="YYYY-MM-DD" value="{% trans "Today" %}" />
+<input class="form-control" type="text" data-field-name="{{ field.name }}"
+       data-date-display-format="{{ request.instance.short_date_format|moment_format }}"
+       data-date-serialize-format="YYYY-MM-DD" value="{% trans "Today" %}" />
 {% else %}
 <input type="text" data-field-name="{{ field.name }}" />
 {% endif %}

--- a/opentreemap/treemap/templatetags/dateformat.py
+++ b/opentreemap/treemap/templatetags/dateformat.py
@@ -23,3 +23,28 @@ def datepicker_format(date_format):
     date_format = date_format.replace('Y', 'yyyy')
 
     return date_format
+
+
+@register.filter
+@stringfilter
+def moment_format(date_format):
+    """
+    Converts a Django DATE_FORMAT to the one specified at:
+    http://momentjs.com/docs/#/displaying/
+    """
+    conv_dict = {
+        'j': 'D',
+        'd': 'DD',
+        'D': 'ddd',
+        'l': 'dddd',
+        'n': 'M',
+        'm': 'MM',
+        'N': 'MMM',
+        'M': 'MMM',
+        'F': 'MMMM',
+        'E': 'MMMM',  # Locale text month,
+        'y': 'YY',
+        'Y': 'YYYY'
+    }
+    moment_format = ''.join(conv_dict.get(ch, ch) for ch in date_format)
+    return moment_format

